internal/column-helpers/columnhelpers.go

```diff
@@ -0,0 +1,307 @@
+	// environment like local-gadget
+	SetColumnFilters(...columns.ColumnFilter)
+
+	// SetSorting sets what sorting should be applied when calling SortEntries() // TODO
+	SetSorting([]string) error
+
+	// SetFilters sets which filter to apply before emitting events downstream
+	SetFilters([]string) error
+
+	// EventHandlerFunc returns a function that accepts an instance of type *T and pushes it downstream after applying
+	// enrichers and filters
+	EventHandlerFunc(enrichers ...func(any)) any
+	EventHandlerFuncArray(enrichers ...func(any)) any
+	EventHandlerFuncSnapshot(key string, enrichers ...func(any)) any
+
+	// JSONHandlerFunc returns a function that accepts a JSON encoded event, unmarshal it into *T and pushes it
+	// downstream after applying enrichers and filters
+	JSONHandlerFunc(enrichers ...func(any)) func([]byte)
+	JSONHandlerFuncArray(enrichers ...func(any)) func([]byte)
+
+	// SetEventCallback sets the downstream callback
+	SetEventCallback(eventCallback any)
+	SetEventCallbackArray(eventCallback any)
+
+	// SetErrorCallback sets the function to use whenever errors occur in this library or are received in-band
+	SetErrorCallback(errorCallback ErrorCallback)
+
+	// EnableSnapshots initializes the snapshot collector, which is able to aggregate snapshots from several sources
+	// and can return (optionally cached) results on demand; used for top gadgets
+	EnableSnapshots(ctx context.Context, t time.Duration, ttl int)
+}
+
+type ColumnHelpers[T any] struct {
+	columns            *columns.Columns[T]
+	sortBy             []string
+	filters            []string
+	filterSpec         *filter.FilterSpec[T] // TODO: filter collection(!)
+	eventCallback      func(*T)
+	eventCallbackArray func([]*T)
+	errorCallback      ErrorCallback
+	snapshotCombiner   *snapshotcombiner.SnapshotCombiner[T]
+	enableSnapshots    bool
+	columnFilters      []columns.ColumnFilter
+}
+
+func NewColumnHelpers[T any](columns *columns.Columns[T]) Columns {
+	ch := &ColumnHelpers[T]{
+		columns: columns,
+	}
+	return ch
+}
+
+func (ch *ColumnHelpers[T]) EnableSnapshots(ctx context.Context, interval time.Duration, ttl int) {
+	if ch.eventCallbackArray == nil {
+		panic("EnableSnapshots needs EventCallbackArray set")
+	}
+	ch.snapshotCombiner = snapshotcombiner.NewSnapshotCombiner[T](ttl)
+	ch.enableSnapshots = true
+	go func() {
+		ticker := time.NewTicker(interval)
+		for {
+			select {
+			case <-ticker.C:
+				out, _ := ch.snapshotCombiner.GetSnapshots()
+				ch.eventCallbackArray(out)
+			case <-ctx.Done():
+			}
+		}
+	}()
+}
+
+func (ch *ColumnHelpers[T]) SetColumnFilters(filters ...columns.ColumnFilter) {
+	ch.columnFilters = filters
+}
+
+func (ch *ColumnHelpers[T]) SetErrorCallback(errorCallback ErrorCallback) {
+	ch.errorCallback = errorCallback
+}
+
+func (ch *ColumnHelpers[T]) SetEventCallback(eventCallback any) {
+	// Check, whether the caller wants a generic callback function (e.g. to print JSON)
```

Comment:
Can you please explain a bit more what you mean?

There's two reasons mainly behind having two of those functions/callbacks:
* gadgets return arrays of *T in snapshots and top categories - and we need them exactly in those batches
* you sadly can't just cast from []any to []*T without that being costly

Reply:
Hmm, that wouldn't work, because `ch.eventCallback` is `func(*T)`, while `ch.eventCallbackArray` is `func([]*T)`. We could of course use `any` for the callback parameter as well, but I think that would make it less readable, typechecking and casting twice.

But using a switch on the type might shorten things a bit and make it more readable.

Reply:
Hehe, yeah :) But I actually found another solution to move it into one function now.